x-pack/plugins/alerting/server/rules_client/rules_client.ts

```diff
@@ -1477,6 +1478,78 @@ export class RulesClient {
     }
   }
 
+  public async snooze({
+    id,
+    snoozeEndTime,
+  }: {
+    id: string;
+    snoozeEndTime?: string;
+  }): Promise<void> {
+    return await retryIfConflicts(
+      this.logger,
+      `rulesClient.snooze('${id}', ${snoozeEndTime})`,
+      async () => await this.snoozeWithOCC({ id, snoozeEndTime })
+    );
+  }
+
+  private async snoozeWithOCC({ id, snoozeEndTime }: { id: string; snoozeEndTime?: string }) {
+    const { attributes, version } = await this.unsecuredSavedObjectsClient.get<RawRule>(
+      'alert',
+      id
+    );
+
+    try {
+      await this.authorization.ensureAuthorized({
+        ruleTypeId: attributes.alertTypeId,
+        consumer: attributes.consumer,
+        operation: WriteOperations.MuteAll,
+        entity: AlertingAuthorizationEntity.Rule,
+      });
+
+      if (attributes.actions.length) {
+        await this.actionsAuthorization.ensureAuthorized('execute');
+      }
+    } catch (error) {
+      this.auditLogger?.log(
+        ruleAuditEvent({
+          action: RuleAuditAction.SNOOZE,
+          savedObject: { type: 'alert', id },
+          error,
+        })
+      );
+      throw error;
+    }
+
+    this.auditLogger?.log(
+      ruleAuditEvent({
+        action: RuleAuditAction.SNOOZE,
+        outcome: 'unknown',
+        savedObject: { type: 'alert', id },
+      })
+    );
+
+    this.ruleTypeRegistry.ensureRuleTypeEnabled(attributes.alertTypeId);
+
+    // If no snoozeEndTime is set, instead mute all
+    const newAttrs = snoozeEndTime
+      ? { snoozeEndTime: new Date(snoozeEndTime).toISOString() }
```

Comment:
There is some business logic that we need to go through like if muteAll is set to true, are we overriding the mute_all meaning, we will set muteAll to false or do we we throw an error that you can not snooze a rule that it is muted.

I think that I prefer to throw an error

Reply:
thinking more about it, maybe we throw an error and add a new query parameters to force the overriding. What do you think?

Reply:
I think we need to see all the options for the values for `snooze_end_time` and `mute_all`, and figure out what has to change when the snooze/mute_all/unmute_all APIs are called with the relevant values, given any possible legal state they could be in within the SO.  Not completely clear to me.

Reply:
> maybe we throw an error and add a new query parameters to force the overriding. What do you think?

Throwing an error seems like not great UX, or will complicate the UX.  Like, you have `mute_all: true` and set a snooze, the error is "You can't mute all and snooze at the same time".  So force the user to unmute_all before proceeding.  

Likewise, an "override" param seems not all that useful - wouldn't we just use it all the time?  Would the "non-overridden" version throw an error if mute_all was true, but the override version wouldn't?  Or what would be the decision tree to get you to the point of using it or not using it.